Importing libraries

In [18]:
import os
import glob
import pandas as pd 
import csv
import collections
import pprint

We display the preprocessed .csv files

In [4]:
data_directory = '../Ivan Compagnucci/EDA-EnterpriseDataAnalytics-project/Data/Merged data/'
raw_files = os.listdir(data_directory)

raw_files

['environmental.csv',
 'luxmeter2018.csv',
 'luxmeter2019.csv',
 'powerelectricity.csv',
 'presence2018.csv',
 'presence2019.csv',
 'reedsensor.csv']

<h3>Now we will verify the integrity of the data of each .csv file</h3> 
<h4 style="color:red"> Environmental.csv </h4>

In [15]:
df = pd.read_csv("../Ivan Compagnucci/EDA-EnterpriseDataAnalytics-project/Data/Merged data/environmental.csv")
df.info()
singleDataNull = sum(df.isnull().values.ravel())
singleRowContainingNull = sum([True for idx,row in df.iterrows() if any(row.isnull())])

print('We have',singleDataNull,'null values in',singleRowContainingNull,'rows')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48173 entries, 0 to 48172
Data columns (total 17 columns):
idsensor                     48173 non-null object
sensortype                   48173 non-null object
srtDate                      48173 non-null object
absolute_humidity            48173 non-null float64
back_front                   48173 non-null float64
battery_level                48173 non-null float64
device_temperature           48044 non-null float64
dew_point                    48044 non-null float64
environmental_temperature    48044 non-null float64
movement_level               48044 non-null float64
orientation                  48044 non-null float64
position                     48044 non-null float64
posture_tilt                 48044 non-null float64
pressure                     48044 non-null float64
regular                      48044 non-null float64
relative_humidity            48044 non-null float64
sequence_number              48044 non-null float64
dtypes: fl

Due to the lack of device data in 129 records, I decided to exclude these rows.
Since in total only 129 lines out of 48173 contain a null value, we can proceed with the elimination of the latter
then saving the .csv file

In [23]:
try:
    df.dropna(inplace=True)
    df.to_csv('../Ivan Compagnucci/EDA-EnterpriseDataAnalytics-project/Data/Cleaned data/environmental.csv')
except OSError:
    print ("Creation of the file failed")
else:
    print("environmental.csv successfully saved!")




environmental.csv successfully saved!


<h4 style="color:red"> Powerelectricity.csv </h4>

In [25]:
df = pd.read_csv("../Ivan Compagnucci/EDA-EnterpriseDataAnalytics-project/Data/Merged data/powerelectricity.csv")
df.info()
singleDataNull = sum(df.isnull().values.ravel())
singleRowContainingNull = sum([True for idx,row in df.iterrows() if any(row.isnull())])

print('We have',singleDataNull,'null values in',singleRowContainingNull,'rows')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 166070 entries, 0 to 166069
Data columns (total 5 columns):
idsensor          166070 non-null object
sensortype        166070 non-null object
srtDate           166070 non-null object
electric_power    166070 non-null float64
rms_current       166070 non-null float64
dtypes: float64(2), object(3)
memory usage: 6.3+ MB
We have 0 null values in 0 rows


No null value seems to be present here. We can save the file without any operation

In [26]:
try:
    df.to_csv('../Ivan Compagnucci/EDA-EnterpriseDataAnalytics-project/Data/Cleaned data/powerelectricity.csv')
except OSError:
    print ("Creation of the file failed")
else:
    print("powerelectricity.csv successfully saved!")


powerelectricity.csv successfully saved!


<h4 style="color:red"> Reedsensor.csv </h4>

In [44]:
df = pd.read_csv("../Ivan Compagnucci/EDA-EnterpriseDataAnalytics-project/Data/Merged data/reedsensor.csv")
df.info()
singleDataNull = sum(df.isnull().values.ravel())
singleRowContainingNull = sum([True for idx,row in df.iterrows() if any(row.isnull())])

print('We have',singleDataNull,'null values in',singleRowContainingNull,'rows')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156958 entries, 0 to 156957
Data columns (total 7 columns):
idsensor              156958 non-null object
sensortype            156958 non-null object
srtDate               156958 non-null object
battery_level         156958 non-null float64
device_temperature    156958 non-null float64
sequence_number       156958 non-null float64
status                156919 non-null float64
dtypes: float64(4), object(3)
memory usage: 8.4+ MB
We have 39 null values in 39 rows


There seems to be a problem in 39 "status" column values, we see in detail:

In [48]:
df_null_status = df[df['status'].isnull()]
df_null_status.head()

idsensor  sensortype                        srtDate  \
14     jzp://edv#0126.0000  ReedSensor  2018-03-01T01:11:11.397+01:00   
4539   jzp://edv#0126.0000  ReedSensor  2018-03-03T13:10:29.557+01:00   
8709   jzp://edv#0126.0000  ReedSensor  2018-03-06T01:11:40.630+01:00   
13447  jzp://edv#0126.0000  ReedSensor  2018-03-08T13:10:47.485+01:00   
18173  jzp://edv#0126.0000  ReedSensor  2018-03-11T01:11:45.825+01:00   

       battery_level  device_temperature  sequence_number  status  
14              3.55                13.0              0.0     NaN  
4539            3.55                12.0              0.0     NaN  
8709            3.55                12.0              0.0     NaN  
13447           3.55                14.0              0.0     NaN  
18173           3.55                13.0              0.0     NaN

Now let's look at the behavior of the "status" value when it is not null

In [51]:
df.groupby('status').agg(['count']).stack()

idsensor  sensortype  srtDate  battery_level  \
status                                                       
0.0    count    152979      152979   152979         152979   
1.0    count      3940        3940     3940           3940   

              device_temperature  sequence_number  
status                                             
0.0    count              152979           152979  
1.0    count                3940             3940

In addition to the 39 null values, there are 3940 "status" values ​​set to 1, so this value should indicate something specific and therefore we cannot set the 39 null values ​​to 0 to solve the problem. 
Then we will save the new file with the null values ​​setted with a marker that we will indicate with "*"

In [56]:
df.fillna("*",inplace=True)
df.info()
df.groupby('status').agg(['count']).stack()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 156958 entries, 0 to 156957
Data columns (total 7 columns):
idsensor              156958 non-null object
sensortype            156958 non-null object
srtDate               156958 non-null object
battery_level         156958 non-null float64
device_temperature    156958 non-null float64
sequence_number       156958 non-null float64
status                156958 non-null object
dtypes: float64(3), object(4)
memory usage: 8.4+ MB


idsensor  sensortype  srtDate  battery_level  \
status                                                       
0.0    count    152979      152979   152979         152979   
1.0    count      3940        3940     3940           3940   
*      count        39          39       39             39   

              device_temperature  sequence_number  
status                                             
0.0    count              152979           152979  
1.0    count                3940             3940  
*      count                  39               39

Now let's save the file cleaned

In [57]:
try:
    df.to_csv('../Ivan Compagnucci/EDA-EnterpriseDataAnalytics-project/Data/Cleaned data/reedsensor.csv')
except OSError:
    print ("Creation of the file failed")
else:
    print("powerelectricity.csv successfully saved!")

powerelectricity.csv successfully saved!


In [ ]:
<h4 style="color:red"> luxmeter2018.csv </h4> and 

It seems that all status values ​​are set to 0 except for 39 rows